File: src/stores/todos.ts
import { defineStore } from 'pinia'
import { invoke } from '@/lib/tauri'
import { ref } from 'vue'

/**
 * Global "quick note" todo list, persisted in the app SQLite DB (todos table)
 * via Tauri commands. Array order mirrors the `position` column — drag-and-drop
 * in the view reorders locally then persists the new id order.
 *
 * Mutations apply optimistically for a snappy UI and refetch on error to resync.
 */
export interface Todo {
  id: string
  text: string
  done: boolean
  position: number
  created_at: string
  /** Capture context: the project this was jotted down in, when any. */
  project_id: string | null
  /** Capture context: the AI run it came from, for the backlink. */
  run_id: string | null
}

export const useTodosStore = defineStore('todos', () => {
  const todos = ref<Todo[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function fetchTodos() {
    loading.value = true
    error.value = null
    try {
      todos.value = await invoke<Todo[]>('list_todos')
    } catch (e) {
      error.value = String(e)
    } finally {
      loading.value = false
    }
  }

  async function add(text: string, projectId?: string | null, runId?: string | null) {
    const trimmed = text.trim()
    if (!trimmed) return
    try {
      const todo = await invoke<Todo>('add_todo', {
        text: trimmed,
        projectId: projectId ?? null,
        // A run backlink only means anything alongside its project.
        runId: projectId ? runId ?? null : null,
      })
      todos.value.unshift(todo)
      return todo
    } catch (e) {
      error.value = String(e)
      await fetchTodos()
    }
  }

  /** Link or unlink a todo from a project (clearing it drops the run backlink). */
  async function setProject(id: string, projectId: string | null) {
    const todo = todos.value.find(t => t.id === id)
    const prev = todo ? { projectId: todo.project_id, runId: todo.run_id } : null
    if (todo) {
      todo.project_id = projectId // optimistic
      if (!projectId) todo.run_id = null
    }
    try {
      await invoke('set_todo_project', { id, projectId })
    } catch (e) {
      error.value = String(e)
      if (todo && prev) {
        todo.project_id = prev.projectId
        todo.run_id = prev.runId
      }
    }
  }

  async function toggle(id: string) {
    const t = todos.value.find(t => t.id === id)
    if (!t) return
    t.done = !t.done // optimistic
    try {
      await invoke('toggle_todo', { id })
    } catch (e) {
      error.value = String(e)
      await fetchTodos()
    }
  }

  async function update(id: string, text: string) {
    const trimmed = text.trim()
    if (!trimmed) {
      await remove(id)
      return
    }
    const t = todos.value.find(t => t.id === id)
    const prev = t?.text
    if (t) t.text = trimmed // optimistic
    try {
      await invoke('update_todo', { id, text: trimmed })
    } catch (e) {
      error.value = String(e)
      if (t && prev !== undefined) t.text = prev
    }
  }

  async function remove(id: string) {
    const i = todos.value.findIndex(t => t.id === id)
    if (i === -1) return
    const [removed] = todos.value.splice(i, 1) // optimistic
    try {
      await invoke('delete_todo', { id })
    } catch (e) {
      error.value = String(e)
      todos.value.splice(i, 0, removed)
    }
  }

  /** Bulk-delete todos by id (optimistic), refetching to resync on error. */
  async function removeMany(ids: string[]) {
    if (ids.length === 0) return
    const target = new Set(ids)
    const snapshot = todos.value.slice()
    todos.value = todos.value.filter(t => !target.has(t.id)) // optimistic
    try {
      await invoke('delete_todos', { ids })
    } catch (e) {
      error.value = String(e)
      todos.value = snapshot
    }
  }

  async function clearCompleted() {
    const snapshot = todos.value.slice()
    todos.value = todos.value.filter(t => !t.done) // optimistic
    try {
      await invoke('clear_completed_todos')
    } catch (e) {
      error.value = String(e)
      todos.value = snapshot
    }
  }

  /** Move the todo at `from` to `to` (drag-and-drop reorder), then persist. */
  async function reorder(from: number, to: number) {
    if (from === to) return
    if (from < 0 || from >= todos.value.length) return
    if (to < 0 || to >= todos.value.length) return
    const snapshot = todos.value.slice()
    const [moved] = todos.value.splice(from, 1)
    todos.value.splice(to, 0, moved)
    try {
      await invoke('reorder_todos', { ids: todos.value.map(t => t.id) })
    } catch (e) {
      error.value = String(e)
      todos.value = snapshot
    }
  }

  return { todos, loading, error, fetchTodos, add, setProject, toggle, update, remove, removeMany, clearCompleted, reorder }
})

File: src-tauri/src/commands/todos.rs
use crate::db::Db;
use serde::Serialize;
use sqlx::Row;
use tauri::State;
use uuid::Uuid;

/// A single quick-note todo. `position` orders the list (ascending: lower value
/// sits higher / higher priority). Drag-and-drop reorder rewrites positions.
///
/// `project_id` / `run_id` are the capture context: set when the todo was jotted
/// down from inside a project's run workspace, so the list can filter by project
/// and link back to the conversation it came from. Both are nullable and
/// unenforced — a globally captured todo has neither.
#[derive(Debug, Serialize, Clone)]
pub struct Todo {
    pub id: String,
    pub text: String,
    pub done: bool,
    pub position: i64,
    pub created_at: String,
    pub project_id: Option<String>,
    pub run_id: Option<String>,
}

fn row_to_todo(row: &sqlx::sqlite::SqliteRow) -> Todo {
    Todo {
        id: row.get("id"),
        text: row.get("text"),
        done: row.get("done"),
        position: row.get("position"),
        created_at: row.get("created_at"),
        project_id: row.get("project_id"),
        run_id: row.get("run_id"),
    }
}

#[tauri::command]
pub async fn list_todos(db: State<'_, Db>) -> Result<Vec<Todo>, String> {
    let rows = sqlx::query(
        "SELECT id, text, done, position, created_at, project_id, run_id FROM todos \
         ORDER BY position ASC, created_at DESC",
    )
    .fetch_all(db.inner())
    .await
    .map_err(|e| e.to_string())?;
    Ok(rows.iter().map(row_to_todo).collect())
}

#[tauri::command]
pub async fn add_todo(
    db: State<'_, Db>,
    text: String,
    project_id: Option<String>,
    run_id: Option<String>,
) -> Result<Todo, String> {
    let text = text.trim().to_string();
    if text.is_empty() {
        return Err("text is required".into());
    }
    let id = Uuid::new_v4().to_string();
    let created_at = chrono::Utc::now().to_rfc3339();
    // New todos go to the top (highest priority) by taking a position below the
    // current minimum.
    let min_pos: Option<i64> = sqlx::query("SELECT MIN(position) AS min_pos FROM todos")
        .fetch_one(db.inner())
        .await
        .map_err(|e| e.to_string())?
        .get("min_pos");
    let position = min_pos.unwrap_or(0) - 1;

    sqlx::query(
        "INSERT INTO todos (id, text, done, position, created_at, project_id, run_id) \
         VALUES (?, ?, 0, ?, ?, ?, ?)",
    )
    .bind(&id)
    .bind(&text)
    .bind(position)
    .bind(&created_at)
    .bind(&project_id)
    .bind(&run_id)
    .execute(db.inner())
    .await
    .map_err(|e| e.to_string())?;

    Ok(Todo {
        id,
        text,
        done: false,
        position,
        created_at,
        project_id,
        run_id,
    })
}

/// Link or unlink a todo from a project. Pass `None` to clear the link; the run
/// backlink is cleared with it, since a run only makes sense inside its project.
#[tauri::command]
pub async fn set_todo_project(
    db: State<'_, Db>,
    id: String,
    project_id: Option<String>,
) -> Result<(), String> {
    if project_id.is_none() {
        sqlx::query("UPDATE todos SET project_id = NULL, run_id = NULL WHERE id = ?")
            .bind(&id)
            .execute(db.inner())
            .await
            .map_err(|e| e.to_string())?;
        return Ok(());
    }
    sqlx::query("UPDATE todos SET project_id = ? WHERE id = ?")
        .bind(&project_id)
        .bind(&id)
        .execute(db.inner())
        .await
        .map_err(|e| e.to_string())?;
    Ok(())
}

#[tauri::command]
pub async fn toggle_todo(db: State<'_, Db>, id: String) -> Result<(), String> {
    sqlx::query("UPDATE todos SET done = 1 - done WHERE id = ?")
        .bind(&id)
        .execute(db.inner())
        .await
        .map_err(|e| e.to_string())?;
    Ok(())
}

#[tauri::command]
pub async fn update_todo(db: State<'_, Db>, id: String, text: String) -> Result<(), String> {
    let text = text.trim().to_string();
    if text.is_empty() {
        return Err("text is required".into());
    }
    sqlx::query("UPDATE todos SET text = ? WHERE id = ?")
        .bind(&text)
        .bind(&id)
        .execute(db.inner())
        .await
        .map_err(|e| e.to_string())?;
    Ok(())
}

#[tauri::command]
pub async fn delete_todo(db: State<'_, Db>, id: String) -> Result<(), String> {
    sqlx::query("DELETE FROM todos WHERE id = ?")
        .bind(&id)
        .execute(db.inner())
        .await
        .map_err(|e| e.to_string())?;
    Ok(())
}

/// Bulk-delete todos by id in a single transaction, so a failure mid-way rolls
/// back the whole batch instead of leaving a partial deletion.
#[tauri::command]
pub async fn delete_todos(db: State<'_, Db>, ids: Vec<String>) -> Result<(), String> {
    if ids.is_empty() {
        return Ok(());
    }
    let mut tx = db.inner().begin().await.map_err(|e| e.to_string())?;
    for id in &ids {
        sqlx::query("DELETE FROM todos WHERE id = ?")
            .bind(id)
            .execute(&mut *tx)
            .await
            .map_err(|e| e.to_string())?;
    }
    tx.commit().await.map_err(|e| e.to_string())?;
    Ok(())
}

#[tauri::command]
pub async fn clear_completed_todos(db: State<'_, Db>) -> Result<(), String> {
    sqlx::query("DELETE FROM todos WHERE done = 1")
        .execute(db.inner())
        .await
        .map_err(|e| e.to_string())?;
    Ok(())
}

/// Persist a new order. The frontend sends the full list of ids top-to-bottom;
/// each row's position becomes its index, so the ordering is stable regardless
/// of prior position values. Done in one transaction to avoid partial reorders.
#[tauri::command]
pub async fn reorder_todos(db: State<'_, Db>, ids: Vec<String>) -> Result<(), String> {
    let mut tx = db.inner().begin().await.map_err(|e| e.to_string())?;
    for (index, id) in ids.iter().enumerate() {
        sqlx::query("UPDATE todos SET position = ? WHERE id = ?")
            .bind(index as i64)
            .bind(id)
            .execute(&mut *tx)
            .await
            .map_err(|e| e.to_string())?;
    }
    tx.commit().await.map_err(|e| e.to_string())?;
    Ok(())
}
